torch/csrc/autograd/engine.cpp

```diff
@@ -1,342 +1,155 @@
-  THPUtils_assert(PyTuple_Check(grad_variables), "variables argument is "
-      "expected to be a tuple, but got %s", THPUtils_typename(grad_variables));
-
-  Py_ssize_t num_variables = PyTuple_GET_SIZE(variables);
-  Py_ssize_t num_gradients = PyTuple_GET_SIZE(grad_variables);
-  THPUtils_assert(num_variables == num_gradients, "got %ld variables and %ld "
-      "gradients", num_variables, num_gradients);
-
+auto Engine::backward(const variable_list& variables,
+                      tensor_list& grad_variables,
+                      bool retain_variables) -> void {
+  function_queue creators;
   ready_queue_type ready;
-  std::unordered_map<THPFunction *, grad_buffer_type> not_ready;
-  dependencies_type dependencies;
-  buffer_set_type need_copy;
 
   bool did_leaf_backward = false;
-  std::vector<THPFunction*> creators;
-  for (int i = 0; i < num_variables; i++) {
-    THPVariable *variable = (THPVariable*)PyTuple_GET_ITEM(variables, i);
-    PyObject *grad = PyTuple_GET_ITEM(grad_variables, i);
-    THPUtils_assert(THPVariable_Check((PyObject*)variable), "element %d of variables "
-        "tuple is not a Variable", i);
-    // If someone calls .backward() on a leaf, it's simple...
-    if (variable->creator == NULL) {
-      if (variable->requires_grad) {
-        THPObjectPtr result = PyObject_CallMethod((PyObject*)variable,
-                "_do_backward", "(O)O", grad, retain_variables_obj);
-        if (!result) return NULL;
+  int size = variables.size();
+  for (int i = 0; i < size; ++i) {
+    auto& var = variables[i];
+    auto& grad = grad_variables[i];
+    if (!var->creator) {
+      // If someone calls .backward() on a leaf, it's simple...
+      if (var->requires_grad) {
+        var->backward(*grad);
         did_leaf_backward = true;
       }
-      continue;
-    }
-    THPFunction *creator = (THPFunction*)variable->creator;
-    creators.push_back(creator);
-    // Initialize the queue
-    if (creator->requires_grad) {
-      grad_buffer_type buf(next_buf_id++, creator->num_outputs);
-      Py_INCREF(grad);
-      buf[variable->output_nr] = grad;
-      ready.emplace_front(creator, std::move(buf));
+    } else {
+      creators.push_back(var->creator);
+      if (var->creator->requires_grad) {
+        GradBuffer buf(var->creator->num_outputs);
+        buf.addGrad(var->output_nr, Variable::of(std::move(grad)));
+        ready.emplace_front(var->creator, std::move(buf));
+      }
     }
   }
 
-  THPEngine_compute_dependencies(std::move(creators), dependencies, ready);
+  auto dependencies = compute_dependencies(creators, ready);
 
-  THPUtils_assert(did_leaf_backward || ready.size() > 0, "there are no graph "
-      "nodes that require computing gradients");
+  if (!did_leaf_backward && ready.size() == 0) {
+    throw std::runtime_error(
+        "there are no graph nodes that require computing gradients");
+  }
 
+  std::unordered_map<Function*, GradBuffer> not_ready;
```

Comment:
It's moved (not copied), so the size of the vector isn't an issue. I'll delete the copy constructor to make this more clear.